agent/agent_endpoint.go

```diff
@@ -1518,6 +1520,13 @@ func (s *HTTPHandlers) AgentConnectCALeafCert(resp http.ResponseWriter, req *htt
 	args.MaxQueryTime = qOpts.MaxQueryTime
 	args.Token = qOpts.Token
 
+	// We don't want non-blocking queries to return expired leaf certs
+	// or leaf certs not valid under the current CA. So always revalidate
+	// the leaf cert on non-blocking queries
+	if args.MinQueryIndex == 0 {
+		args.MustRevalidate = true
+	}
```

Comment:
I think this logic might be better in `ConnectCALeafRequest.CacheInfo()`. I believe that method should have access to all the same data.

By moving it there we can be sure that all users of the cache do the right thing, instead of only this one HTTP API endpoint.

Reply:
I think I'm going to walk this back.

Here's what I ran into -- there's a non trivial amount of ConnectCALeaf testing that would need to change as a result of always passing in `MustRevalidate` assignment logic in the CacheInfo(). It kinda feels like this behavior is "breaking" with previous one. It seems, ergonomically, this func was designed more as a "getter".

Reply:
hmm, that is unfortunate. I think those tests must not be accurately testing the production behaviour very well, but that is cleanup we can leave for another time I think. Thanks for trying.

Maybe leave a TODO here instead? Not critical